src/server/game/Server/WorldSocket.cpp

```diff
@@ -153,58 +187,140 @@ const std::string& WorldSocket::GetRemoteAddress (void) const
     {
-        buff.Compress(m_Session->GetCompressionStream(), pkt);
-        pkt = &buff;
-    }*/
+        size_t size = pct->size();
+        uint32 bufferSize = deflateBound(m_zstream, pct->size() + SizeOfServerHeader[m_Crypt.IsInitialized()]);
 
-    uint16 opcodeNumber = serverOpcodeTable[pkt->GetOpcode()]->OpcodeNumber;
+        size_t reserved_size = bufferSize + 12;
+        m_CompressedPct.resize(reserved_size);
+        m_CompressedPct.put<uint32>(0, size + 4);
 
-    if (m_Session)
-        SF_LOG_TRACE("network.opcode", "S->C: %s %s", m_Session->GetPlayerInfo().c_str(), GetOpcodeNameForLogging(pkt->GetOpcode(), true).c_str());
+        uint32 adler_1 = adler32(adler32(0x9827D8F1, (Bytef*)&opcodeNumber, 4), pct->contents(), size);
 
-    sScriptMgr->OnPacketSend(this, *pkt);
+        m_CompressedPct.put<uint32>(4, adler_1);
 
-    ServerPktHeader header(!m_Crypt.IsInitialized() ? pkt->size() + 2 : pct.size(), opcodeNumber, &m_Crypt);
+        m_zstream->next_out = (Bytef*)(m_CompressedPct.contents() + 12);
+        m_zstream->avail_out = bufferSize;
 
-    if (m_OutBuffer->space() >= pkt->size() + header.getHeaderLength() && msg_queue()->is_empty())
+        m_zstream->next_in = (Bytef*)&opcodeNumber;
+        m_zstream->avail_in = 4;
+
+        int z_res = deflate(m_zstream, Z_BLOCK);
+        if (z_res != Z_OK)
+        {
+            SF_LOG_ERROR("network.io", "Can't compress packet opcode (zlib: deflate) Error code: %i (%s, msg: %s)", z_res, zError(z_res), m_zstream->msg);
+            return 0;
+        }
+
+        m_zstream->next_in = (Bytef*)pct->contents();
+        m_zstream->avail_in = pct->size();
+
+        z_res = deflate(m_zstream, Z_SYNC_FLUSH);
+        if (z_res != Z_OK)
+        {
+            SF_LOG_ERROR("network.io", "Can't compress packet data (zlib: deflate) Error code: %i (%s, msg: %s)", z_res, zError(z_res), m_zstream->msg);
+            return 0;
+        }
+
+        uint32 compressedSize = bufferSize - m_zstream->avail_out;;
+        uint32 adler_2 = (uint32)adler32(0x9827D8F1, (const Bytef*)(m_CompressedPct.contents() + 12), compressedSize);
+
+        m_CompressedPct.put<uint32>(8, adler_2);
+
+        if (z_res != Z_OK)
+        {
+            SF_LOG_ERROR("network.io", "Can't compress packet (zlib: deflate) Error code: %i (%s)",z_res,zError(z_res));
+        }
+        else if (m_zstream->avail_in != 0)
+        {
+            SF_LOG_ERROR("network.io", "Can't compress packet (zlib: deflate not greedy)");
+        }
+        else
+        {
+            m_CompressedPct.resize(compressedSize + 12);
+            m_CompressedPct.SetOpcode(SMSG_COMPRESSED_PACKET);
+            pct = &m_CompressedPct;
+        }
+
+        m_zstream->next_in = NULL;
+        m_zstream->next_out = NULL;
+        m_zstream->avail_in = 0;
+        m_zstream->avail_out = 0;
+    }
+
+    ServerPktHeader header(!m_Crypt.IsInitialized() ? pct->size() + 2 : pct->size(), opcodeNumber, &m_Crypt);
```

Comment:
opcodeNumber needs reinitizalize, because we changed opcode to SMSG_COMPRESSED_PACKET, so opcodeNumber is different now.
That's why now is crashes the client.

Reply:
This pull was closed